src/connection.ts

```diff
@@ -655,6 +665,20 @@ export class Connection extends EventEmitter {
       }
     }
 
+    // Allow consumer to choose to fulfill each packet and/or perform other logic before fulfilling
+    if (this.shouldFulfill && incomingAmount.greaterThan(0)) {
+      await this.shouldFulfill(this.connectionId, requestPacket.sequence, incomingAmount).catch(async err => {
+        this.log.debug('application declined to fulfill packet %s:', requestPacket.sequence, err)
+
+        await throwFinalApplicationError()
```

Comment:
Do you think we should allow ILP errors to be thrown out of the shouldFulfill? It would give more flexibility (if some system were waiting to come online or load invoice details then shouldFulfill could get the stream to wait with T00's, for instance), but it also could interact really weirdly with the rest of the STREAM behavior.

Reply:
Possibly. I'm not huge of the application having to understand ILP errors, but it could be nice for the callback to have the flexibility to e.g. indicate the connection should be terminated, or that the client should retry with some backoff. My lean would be to wait until we have a better idea what issues crop up/how this is being deployed before adding that functionality